### ISSUE TITLE
breaking Appsignal.set_error API changes since 2.x

### ISSUE THREAD

**Issue by reneweteling** [appsignal/appsignal-elixir#610]
## Describe the bug

When using set_error and the first attribute != :error, appsignal errors when the `APPSIGNAL_PUSH_API_KEY` is present, and when its omitted all is well. This gave us some issues with a green testsuite and still issues on production. 

The following code:

```elixir
{:current_stacktrace, current_stacktrace} = Process.info(self(), :current_stacktrace)
    Appsignal.set_error("AvidciFetchRecommendationError", "200 - response: [6657]", current_stacktrace)
```

Would trigger this issue, only when Appsignal is enabled, so the enabled API interface behaves differently than the disabled API interface.

```elixir
   The following arguments were given to Exception.format_banner/3:

         # 1
         "AvidciFetchRecommendationError"

         # 2
         "200 - response: [6657]"

         # 3
         [{Process, :info, 2, [file: 'lib/process.ex', line: 766]}, {McCore.AppSignal, :notify, 2, [file: 'lib/mc_core/helpers/appsignal.ex', line: 10]}, {VaasWeb.Embed.EmbedSDKController, :recommendations, 2, [file: 'lib/vaas_web/controllers/embed/embed_sdk_controller.ex', line: 111]}, {VaasWeb.Embed.EmbedSDKController, :action, 2, [file: 'lib/vaas_web/controllers/embed/embed_sdk_controller.ex', line: 1]}, {VaasWeb.Embed.EmbedSDKController, :phoenix_controller_pipeline, 2, [file: 'lib/vaas_web/controllers/embed/embed_sdk_controller.ex', line: 1]}, {Phoenix.Router, :__call__, 2, [file: 'lib/phoenix/router.ex', line: 352]}, {VaasWeb.Endpoint, :plug_builder_call, 2, [file: 'lib/vaas_web/endpoint.ex', line: 1]}, {VaasWeb.Endpoint, :"-call (overridable 3)/2-fun-0-", 3, [file: 'lib/vaas_web/endpoint.ex', line: 3]}, {Appsignal.Instrumentation, :instrument, 1, [file: 'lib/appsignal/instrumentation.ex', line: 9]}, {VaasWeb.Endpoint, :"call (overridable 4)", 2, [file: 'lib/plug/debugger.ex', line: 132]}, {VaasWeb.Endpoint, :"call (overridable 5)", 2, [file: 'lib/vaas_web/endpoint.ex', line: 1]}, {VaasWeb.Endpoint, :call, 2, [file: 'lib/plug/error_handler.ex', line: 65]}, {Phoenix.ConnTest, :dispatch, 5, [file: 'lib/phoenix/test/conn_test.ex', line: 225]}, {VaasWeb.Embed.EmbedSDKControllerTest, :"test EmbedSDK Controller get recommendations from recommender avidci (tests mychannels recommender fallback)", 1, [file: 'test/vaas_web/controllers/embed/embed_sdk_controller_test.exs', line: 104]}, {ExUnit.Runner, :exec_test, 1, [file: 'lib/ex_unit/runner.ex', line: 391]}, {:timer, :tc, 1, [file: 'timer.erl', line: 166]}, {ExUnit.Runner, :"-spawn_test_monitor/4-fun-1-", 4, [file: 'lib/ex_unit/runner.ex', line: 342]}]

     Attempted function clauses (showing 4 out of 4):

         def format_banner(:error, exception, stacktrace)
         def format_banner(:throw, reason, _stacktrace)
         def format_banner(:exit, reason, _stacktrace)
         def format_banner({:EXIT, pid}, reason, _stacktrace)
```

I have changed all our calls to:

```elixir
{:current_stacktrace, current_stacktrace} = Process.info(self(), :current_stacktrace)
    Appsignal.set_error(:error, "AvidciFetchRecommendationError - 200 - response: [6657]", current_stacktrace)
```

But this will probably ruin the title in the Appsignal dashboard. 

Hope to hear from you soon, thanks guys.

**Comment by thijsc** [comment]
It would be good if the new `set_error` supported strings as names too. We'll add that back. Thanks for reporting!

**Comment by jeffkreeftmeijer** [comment]
Hey René! 

We recommend sending an exception through `set_error/2-4`. Although passing a string will work, we have trouble turning that into an actual exception and then correctly categorizing them on our end.

Erlang doesn’t have a data structure to describe exceptions, meaning internally, an `ArgumentError` is simply two atoms; the “kind”, which is `:error`, and the “reason” which is `:badarg`. 

We use Elixir’s `Exception.normalize/3` to convert any error we catch to an actual exception internally:

``` elixir
Exception.normalize(:error, :badarg, [])  
# => %ArgumentError{message: "argument error"}
```

This way, we always have Elixir-style exceptions in the backend, and we can group anything coming straight from Erlang with anything coming from Elixir.

If you pass a string, `Exception.normalize/3` doesn’t know what to do with that, so it converts it to a more general `ErlangError`:

``` elixir
Exception.normalize(:error, "String!", [])
# => %ErlangError{original: "String!"}
```

On our processing side, we use the exception type for grouping your exceptions. An exception like that `ErlangError` would essentially be thrown on a pile with the other exceptions from that same action.

That’s why I’d recommend using a custom error type, and sending that. In your case maybe a custom `FetchRecommendationError` would be a good fit:

``` elixir
defmodule FetchRecommendationError do
  defexception message: "default message"
end
```

Then, if you raise and catch that, you’ll be able to send it over to AppSignal, and have it correctly categorized:

``` elixir
try do
  raise "FetchRecommendationError!"
Catch
  kind, reason -> Appsignal.set_error(kind, rason, __STACKTRACE__)
end
```

Here, we recommend raising an error instead of calling out to `Process.info/2` to make sure the stack trace is created exactly when the error is raised, instead of when it was caught.

## Upcoming in 2.1.0

Your issue prompted us to take another look at how we’ve implemented the `set_error/2-4` helpers in 2.0, and we’re making two improvements to make them a bit more flexible.

One might be helpful for you in particular. [Version 2.1 will allow you to pass an exception without the “kind”](https://github.com/appsignal/appsignal-elixir/pull/620) (as an Elixir exception is always an `:error`):

```
try do
  raise "Exception!"
rescue
  exception -> Appsignal.set_error(exception, __STACKTRACE__)
end
```

**Comment by jeffkreeftmeijer** [comment]
Heads up! We’ve released version 2.1.0 of the `:appsignal` library, which include the improved error helpers described above. To upgrade, please update the dependency on all three libraries in your app. For Phoenix:

```
$ mix deps.update appsignal_phoenix appsignal_plug appsignal
```
